design/configuration-crd.md

```diff
@@ -0,0 +1,210 @@
+Contour will provide a way internally to move to the new CRD and not require users to manually migrate to the new CRD format.
+Contour will provide a new command or external tool (similar to ir2proxy) which will migrate between the specs accordingly. 
+
+## Contour Deployment
+
+A managed version of Contour was made available with the `Contour Operator`.
+Since Contour will manage Envoy instances, the Operator will now manage instances of Contour.
+The details of how an instance of Contour should be deployed within a cluster will be defined in the second CRD named `ContourDeployment`. 
+The `spec.confguration` of this object will be the same struct defined in the `ContourConfiguration`. 
+
+A controller will watch for these objects to be created and take action on them accordingly to make desired state in the cluster match the configuration on the spec. 
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: ContourDeployment
+metadata:
+  name: contour
+spec:
+  replicas: 2
+  nodePlacement:
+    nodeSelector:
+    tolerations:
+  configuration:
+    <same config as above>
+status:
+```
+
+## Processing Logic
+
+Contour will require a new flag (`--contour-config`), which will allow for customizing the name of the `ContourConfiguration` CRD that is it to use.
+It will default to one named `contour`, but could be overridden if desired.
+The ContourConfiguration referenced must also be in the same namespace as Contour is running, it's not valid to reference a configuration from another namespace.
+The current flag `--config-path`/`-c` will continue to point to the Configmap file, but over time could eventually be deprecated and the short code `-c` be used for the CRD location (i.e. `--contour-config`) for simplicity.
+The Contour configuration CRD will still remain optional.
+In its absence, Contour will operate with reasonable defaults.
+Where Contour settings can also be specified with command-line flags, the command-line value takes precedence over the configuration file.
+
+On startup, Contour will look for a `ContourConfiguration` CRD named `contour` or using the value of the flag `--contour-config` in the same namespace which Contour is running in, Contour won't support referencing from a different namespace.
+If the `ContourConfiguration` CRD is not found Contour will start up with reasonable defaults.
+
+Contour will set status on the object informing the user if there are any errors or issues with the config file or that is all fine and processing correctly.
+
+Once Contour begins using a Configuration CRD, it will add a finalizer to it such that if that resource is going to get deleted, Contour is aware of it.
+Should the Configuration CRD be deleted while it is in use, Contour will default back to reasonable defaults and log the issue.
+
+When config in the CRD changes we will gracefully stop the dependent ingress/gateway controllers and restart them with new config, or dynamically update some in-memory data that the controllers use.
```

Comment:
I think this basically means re-running `doServe(...)`, right? Related to my previous comment, I think it's worth discussing whether it's worth writing code to handle this or whether we should simply trigger a pod restart. Personally I'd lean towards the latter for simplicity, unless there's a compelling reason to do the former.

Reply:
I agree to start with the simple bits first before trying to over optimize.

Reply:
yup we'll start with the crash loop, but I think that we can do better. Our doServe() isn't really tested so that's a ton of room for improvement there. I don't feel that killing/restarting internal controllers is impossible for us to figure out. Our goal should be to best serve traffic to users. But can't do that all at first, so let's start here and improve later.

Reply:
Restart approach is of course good starting point to experiment, but I think users would bump into [crashloop & exponential back-off timer problem](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy). Even if single configuration error would not trigger crashloop, a series of attribute changes (and therefore restarts) may look like crashloop to Kubernetes - making re-configuration slower and slower as the timer is increased.